lib/core/mergeConfig.js

```diff
@@ -0,0 +1,44 @@
+'use strict';
+
+var utils = require('../utils');
+
+/**
+ * Config-specific merge-function which creates a new config-object
+ * based on given defaults and instance config.
+ *
+ * @param {Object} defaults Defaults
+ * @param {Object} instanceConfig Instance-specific config
+ * @returns {Object} New object resulting from merging instanceConfig to defaults
+ */
+module.exports = function mergeConfig(defaults, instanceConfig) {
+  // eslint-disable-next-line no-param-reassign
+  instanceConfig = instanceConfig || {};
+  var config = {};
+
+  utils.forEach(['url', 'method', 'params', 'data'], function valueFromInstanceConfig(prop) {
+    config[prop] = instanceConfig[prop];
+  });
+
+  utils.forEach(['headers', 'auth', 'proxy'], function mergeInstanceConfigWithDefaults(prop) {
+    if (typeof instanceConfig[prop] !== 'undefined') {
```

Comment:
I actually was not able to simplify this at all. If I got rid of explicitly checking for undefined and just did `||`, the tests would fail. I assume something somewhere relies on `null` values and that broke it. If absolutely necessary, I can try to dig deeper into this later. I do agree that the if-statements are nasty.